File: src/main/java/mc/fbturrets/blocks/TurretHolderBlockEntity.java
package mc.fbturrets.blocks;

import io.netty.buffer.Unpooled;
import mc.fbturrets.main.FBTurrets;
import mc.fbturrets.math.MathHelp;
import net.fabricmc.fabric.api.network.ServerSidePacketRegistry;
import net.fabricmc.fabric.api.server.PlayerStream;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.entity.*;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.entity.decoration.ArmorStandEntity;
import net.minecraft.entity.decoration.EndCrystalEntity;
import net.minecraft.entity.mob.EndermanEntity;
import net.minecraft.entity.passive.PigEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.projectile.ProjectileEntity;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.util.Tickable;
import net.minecraft.util.hit.HitResult;
import net.minecraft.util.math.Box;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.RayTraceContext;

import java.util.List;
import java.util.stream.Stream;

public class TurretHolderBlockEntity extends BlockEntity implements Tickable {
    private final Box targetBox = new Box(-15, -15, -15, 15, 15, 15);
    private int currentAimTime = 0;
    private Vec3d turretVec3d;

    public float yaw = 0;
    public float pitch = 0;

    public TurretGun gun = FBTurrets.TEST_GUN;
    public Entity target;

    public TurretHolderBlockEntity() {
        super(FBTurrets.TURRET_HOLDER_BLOCK_ENTITY);
    }

    @Override
    public void tick() {
        assert world != null;
        if (!world.isClient) {

            turretVec3d = new Vec3d(pos.getX() + 0.5, pos.getY() + 0.65, pos.getZ() + 0.5);

            currentAimTime++;

            List<Entity> targets = world.getEntities(null, targetBox.offset(pos));
            if (!targets.isEmpty()) {

                targets.removeIf(this::cantBeAttacked);

                for (Entity possible : targets) {
                    if (target == null) {
                        target = possible;
                        currentAimTime = 0;
                        break;
                    }
                }
            }

            if (target != null && !target.removed && target.isAlive()) {
                Vec3d targetPos = new Vec3d(target.getX(), (target.getPos().y + target.getBoundingBox().getYLength() / 2), target.getZ());

                double d = target.getX() - turretVec3d.x;
                double e = (target.getY() + target.getBoundingBox().getYLength() / 2) - turretVec3d.y;
                double f = target.getZ() - turretVec3d.z;
                double g = MathHelper.sqrt(d * d + f * f);

                float target_pitch = MathHelper.wrapDegrees((float) (-(MathHelper.atan2(e, g) * 57.2957763671875D)));
                float target_yaw = MathHelper.wrapDegrees((float) (MathHelper.atan2(f, d) * 57.2957763671875D) - 90.0F);

                pitch = MathHelp.lerpAngle(pitch, target_pitch, 0.4f);
                yaw = MathHelp.lerpAngle(yaw, target_yaw, 0.4f);

                Stream<PlayerEntity> watchingPlayers = PlayerStream.watching(world, pos);
                PacketByteBuf passedData = new PacketByteBuf(Unpooled.buffer());
                passedData.writeBlockPos(pos);
                passedData.writeFloat(pitch);
                passedData.writeFloat(yaw);
                watchingPlayers.forEach(player -> ServerSidePacketRegistry.INSTANCE.sendToPlayer(player, FBTurrets.TURRET_ANGLES_UPDATE, passedData));

                Vec3d offset = turretVec3d.add(MathHelp.getDirectionVector(pitch - 180, yaw).multiply(1.2));

                HitResult.Type result = world.rayTrace(new RayTraceContext(offset, targetPos, RayTraceContext.ShapeType.COLLIDER, RayTraceContext.FluidHandling.ANY, new PigEntity(EntityType.PIG, world))).getType();

                if (currentAimTime >= gun.getAimTime() && (result == HitResult.Type.MISS || result == HitResult.Type.ENTITY)) {
                    currentAimTime = 0;
                    target.damage(DamageSource.GENERIC, gun.getDamage());

                    Stream<PlayerEntity> newWatchingPlayers = PlayerStream.watching(world, pos);
                    PacketByteBuf newPassedData = new PacketByteBuf(Unpooled.buffer());
                    newPassedData.writeBlockPos(pos);
                    newPassedData.writeDouble(d);
                    newPassedData.writeDouble(e);
                    newPassedData.writeDouble(f);
                    newWatchingPlayers.forEach(player -> ServerSidePacketRegistry.INSTANCE.sendToPlayer(player, FBTurrets.TURRET_SHOOT_PARTICLES, newPassedData));
                } else {
                    if (result == HitResult.Type.BLOCK) {
                        target = null;
                    }
                }
            } else {
                target = null;
            }
        } else {
            target = null;
        }
    }

    boolean cantBeAttacked(Entity target) {
        assert world != null;
        if (target != null) {
            if (!target.isAlive() || (
                    target instanceof PlayerEntity ||
                            target instanceof ItemEntity ||
                            target instanceof ExperienceOrbEntity ||
                            target instanceof ProjectileEntity ||
                            target instanceof ArmorStandEntity ||
                            target instanceof TntEntity ||
                            target instanceof EndCrystalEntity ||
                            target instanceof EndermanEntity ||
                            target instanceof AreaEffectCloudEntity)
            ) {
                return true;
            }

            Vec3d targetPos = new Vec3d(target.getX(), (target.getPos().y + target.getBoundingBox().getYLength() / 2), target.getZ());
            Vec3d dir = targetPos.subtract(turretVec3d).normalize();
            HitResult.Type result = world.rayTrace(new RayTraceContext(turretVec3d.add(dir.multiply(1.2)), targetPos, RayTraceContext.ShapeType.COLLIDER, RayTraceContext.FluidHandling.ANY, new PigEntity(EntityType.PIG, world))).getType();
            return result == HitResult.Type.BLOCK;
        }
        return true;
    }
}

File: src/main/kotlin/mc/fbturrets/blocks/TurretHolderBlockEntity.kt
package mc.fbturrets.blocks

import io.netty.buffer.Unpooled
import mc.fbturrets.main.FBTurrets
import mc.fbturrets.util.MathHelp.absDist
import mc.fbturrets.util.MathHelp.facingDifference
import mc.fbturrets.util.MathHelp.lerpAngle
import mc.fbturrets.util.MathHelp.smallRandom
import net.fabricmc.fabric.api.network.ServerSidePacketRegistry
import net.fabricmc.fabric.api.server.PlayerStream
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityType
import net.minecraft.entity.ItemEntity
import net.minecraft.entity.damage.DamageSource
import net.minecraft.entity.mob.Monster
import net.minecraft.entity.passive.PigEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.Inventories
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.PacketByteBuf
import net.minecraft.screen.NamedScreenHandlerFactory
import net.minecraft.screen.ScreenHandler
import net.minecraft.text.LiteralText
import net.minecraft.text.Text
import net.minecraft.text.TranslatableText
import net.minecraft.util.Identifier
import net.minecraft.util.Tickable
import net.minecraft.util.collection.DefaultedList
import net.minecraft.util.hit.HitResult
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.Vec3d
import net.minecraft.util.registry.Registry
import net.minecraft.world.RayTraceContext

class TurretHolderBlockEntity : BlockEntity(FBTurrets.TURRET_HOLDER_BLOCK_ENTITY), Tickable, ImplementedInventory,
    NamedScreenHandlerFactory {
    private var inventory = DefaultedList.ofSize(4, ItemStack.EMPTY)
    private var currentAimTime = 0
    private var turretVec3d: Vec3d? = null
    private var tickCount = 0

    @JvmField
    var yaw = 0f

    @JvmField
    var pitch = 0f

    @JvmField
    var gun: TurretGun? = null
    private var target: Entity? = null
    override fun tick() {
        if (gun != null && world != null) {
            val targetBox = gun!!.targetBox
            if (!world!!.isClient && gun != null) {
                tickCount++
                if (tickCount % 20 == 0) {
                    tickCount = 0
                    val playersAround = PlayerStream.watching(world, pos)
                    val gunPassedData = PacketByteBuf(Unpooled.buffer())
                    gunPassedData.writeBlockPos(pos)
                    gunPassedData.writeString(gun!!.iD.toString())
                    playersAround.forEach { player: PlayerEntity? ->
                        ServerSidePacketRegistry.INSTANCE.sendToPlayer(
                            player,
                            FBTurrets.TURRET_GUN_UPDATE,
                            gunPassedData
                        )
                    }
                }
                turretVec3d = Vec3d(pos.x + 0.5, pos.y + 0.65, pos.z + 0.5)
                val targets = world!!.getEntities(null, targetBox.offset(pos))
                if (!targets.contains(target)) {
                    target = null
                }
                if (targets.isNotEmpty()) {
                    targets.removeIf { target: Entity? -> cantBeAttacked(target) }
                    for (possible in targets) {
                        if (target == null) {
                            target = possible
                            currentAimTime = 0
                            break
                        }
                    }
                }
                if (target != null && !target!!.removed && target!!.isAlive) {
                    val d = target!!.x - turretVec3d!!.x
                    val e = target!!.y + target!!.boundingBox.yLength / 2 - turretVec3d!!.y
                    val f = target!!.z - turretVec3d!!.z
                    val g = MathHelper.sqrt(d * d + f * f).toDouble()
                    val targetPitch = MathHelper.wrapDegrees((-(MathHelper.atan2(e, g) * 57.2957763671875)).toFloat())
                    val targetYaw =
                        MathHelper.wrapDegrees((MathHelper.atan2(f, d) * 57.2957763671875).toFloat() - 90.0f)
                    val lastPitch = pitch
                    val lastYaw = yaw
                    pitch = lerpAngle(pitch, targetPitch, 0.1f)
                    yaw = lerpAngle(yaw, targetYaw, 0.1f)
                    if (absDist(pitch, lastPitch) >= 0.05 || absDist(yaw, lastYaw) >= 0.05) {
                        val watchingPlayers = PlayerStream.watching(world, pos)
                        val passedData = PacketByteBuf(Unpooled.buffer())
                        passedData.writeBlockPos(pos)
                        passedData.writeFloat(pitch)
                        passedData.writeFloat(yaw)
                        watchingPlayers.forEach { player: PlayerEntity? ->
                            ServerSidePacketRegistry.INSTANCE.sendToPlayer(
                                player,
                                FBTurrets.TURRET_ANGLES_UPDATE,
                                passedData
                            )
                        }
                    }
                    if (facingDifference(pitch, MathHelper.wrapDegrees(yaw), targetPitch, targetYaw) < 3) {
                        currentAimTime++
                        if (currentAimTime >= gun!!.aimTime && !cantBeAttacked(target)) {
                            currentAimTime = 0
                            target!!.damage(DamageSource.GENERIC, gun!!.damage.toFloat())
                            val newWatchingPlayers = PlayerStream.watching(world, pos)
                            val newPassedData = PacketByteBuf(Unpooled.buffer())
                            newPassedData.writeBlockPos(pos)
                            newPassedData.writeDouble(d)
                            newPassedData.writeDouble(e)
                            newPassedData.writeDouble(f)
                            newWatchingPlayers.forEach { player: PlayerEntity? ->
                                ServerSidePacketRegistry.INSTANCE.sendToPlayer(
                                    player,
                                    FBTurrets.TURRET_SHOOT_PARTICLES,
                                    newPassedData
                                )
                            }
                        } else {
                            if (currentAimTime >= gun!!.aimTime) {
                                target = null
                            }
                        }
                    }
                } else {
                    target = null
                }
            } else {
                target = null
            }
        }
    }

    private fun cantBeAttacked(target: Entity?): Boolean {
        if (target != null && world != null) {
            if (!target.isAlive) {
                return true
            }
            if (target !is Monster) {
                return true
            }
            val targetPos = Vec3d(target.x, target.pos.y + target.boundingBox.yLength / 2, target.z)
            val dir = targetPos.subtract(turretVec3d).normalize()
            val result = world!!.rayTrace(
                RayTraceContext(
                    turretVec3d!!.add(dir.multiply(1.2)),
                    targetPos,
                    RayTraceContext.ShapeType.COLLIDER,
                    RayTraceContext.FluidHandling.ANY,
                    PigEntity(EntityType.PIG, world)
                )
            ).type
            return result == HitResult.Type.BLOCK
        }
        return true
    }

    fun dropGun() {
        if (gun != null) {
            val itemEntity =
                ItemEntity(world, pos.x + 0.5, (pos.y + 1).toDouble(), pos.z + 0.5, ItemStack(Registry.ITEM[gun!!.iD]))
            assert(world != null)
            itemEntity.setVelocity(smallRandom(), 0.15, smallRandom())
            world!!.spawnEntity(itemEntity)
        }
    }

    override fun toTag(tag: CompoundTag): CompoundTag {
        if (gun != null) {
            tag.putString("gunID", gun!!.iD.toString())
        } else {
            tag.putString("gunID", "")
        }
        Inventories.toTag(tag, items);
        return super.toTag(tag)
    }

    override fun fromTag(state: BlockState, tag: CompoundTag) {
        val id = tag.getString("gunID")
        if (id.isNotEmpty()) {
            gun = Registry.BLOCK[Identifier(id)] as TurretGun
        }
        Inventories.fromTag(tag,items);
        super.fromTag(state, tag)
    }

    override fun getItems(): DefaultedList<ItemStack> {
        return inventory
    }

    override fun markDirty() {
        super<BlockEntity>.markDirty()
    }

    override fun createMenu(syncId: Int, inv: PlayerInventory?, player: PlayerEntity?): ScreenHandler? {
        TODO("Not yet implemented")
    }

    override fun getDisplayName(): Text {
        return TranslatableText(cachedState.block.translationKey)
    }
}